### ISSUE TITLE
Enable numpydoc validation

### ISSUE THREAD

**Issue by vyasr** [NVIDIA/cudf#11253]
numpydoc supports [docstring validation](https://numpydoc.readthedocs.io/en/latest/validation.html) that we should enable to supplement things like pydocstyle checks. The numpydoc validation will help us with more semantic checking, such as ensuring that docstrings actually align with the function signatures that they are attached to.

**Comment by ShahRishi** [comment]
Hi, I'd like to help out on this. When you say enabling docstring validation, is this the same thing as running it?

**Comment by vyasr** [comment]
Hi @ShahRishi, we'd love your help! It's not quite that simple, although it's close. You'll need to build the documentation with the checks turned on (that's how you run it), but then you will also need to fix any issues that it finds. [Here are the instructions on building the docs](https://github.com/rapidsai/cudf/tree/branch-22.08/docs/cudf). My recommendation would probably be to clone the repository and install the conda environment as it recommends, but instead of building cudf from source just try to install cudf from the `rapidsai-nightly` conda channel so that you get the most up-to-date version of the package without having to build it yourself. Then you should be able to start from step 2 on that page and build the documentation as described. Does that make sense? Let me know if you have any further questions!

**Comment by ShahRishi** [comment]
Hi, just wanted to let you know that I'm still working on this. Sorry for taking so long, this is new for me!

**Comment by ShahRishi** [comment]
Hi, I think I realize the problem that I'm having. I am working on a 3060ti GPU, which is not listed as a RAPIDS compatible GPU. Am I able to build the documentation with this GPU?

**Comment by vyasr** [comment]
Hmm I don't think that should be a problem. We only require a Pascal architecture, and the 3060ti is Ampere. What errors are you running into? Did you actually build cudf before building the documentation? cudf must be installed before building the docs.

**Comment by vyasr** [comment]
> Hi, just wanted to let you know that I'm still working on this. Sorry for taking so long, this is new for me!

No worries! Please feel free to keep asking questions.

**Comment by ShahRishi** [comment]
Just wanted to let you know I haven't abandoned this and am still working!

**Comment by shwina** [comment]
No problem! Please let us know if you need help with anything. You can also find us on [Slack](https://join.slack.com/t/rapids-goai/shared_invite/zt-trnsul8g-Sblci8dk6dIoEeGpoFcFOQ).

**Comment by vyasr** [comment]
@ShahRishi were you able to resolve the build issues? Hope that isn't still holding you up.

**Comment by ShahRishi** [comment]
Hi, I apologize for the delay. I've had to put this on hold as I've been studying for the MCAT exam through August. I am still working and am learning a lot. I know this is taking me far longer than it should, but I am determined to figure this out. Thank you so much for your patience.

**Comment by shwina** [comment]
Hi @ShahRishi - it's no problem! We'll keep this open and assigned to you. Feel free to pick up wherever you left off at any time or just open a new PR if that's easier.

**Comment by ShahRishi** [comment]
Hello, thank you again for your patience. I have successfully added the validation checks and rebuilt the docs. I realized that the problem I was having was that I had an outdated version of cmake. After running the validation checks, several thousands of warnings appeared. Should I fix all of these warnings? If not, among the full mapping of validation checks are there only certain ones I should address? Thank you!

**Comment by vyasr** [comment]
Hi @ShahRishi, I'm glad that you were able to get this working! Could you post a sample of some of the warnings that you are seeing? Since this is the first time that we're enabling this, I think what we will want to do is make a sequence of PRs where each one just fixes all instances of one or two errors. That will let us figure out which warnings are worth fixing and which ones (if any) we want to ignore. If you can post some examples on this issue, we can evaluate those warnings a few at a time.

**Comment by ShahRishi** [comment]
Hi @vyasr! Thank you for the feedback. I am unfortunately unable to access my computer right now, but I will paste all of the warnings delivered and post them after 9pm PST tonight! Thank you again!

**Comment by vyasr** [comment]
Great, thank you! If you want to paste _all_ of the warnings, please paste them into a text file and attach it here rather than copying directly, otherwise it will be very difficult to parse through them on the Github UI.

**Comment by ShahRishi** [comment]
Hi! I have uploaded the warnings and added a file below. I have also provided a list of the types of errors from the documentation
`ERROR_MSGS = {
    "GL01": "Docstring text (summary) should start in the line immediately "
    "after the opening quotes (not in the same line, or leaving a "
    "blank line in between)",
    "GL02": "Closing quotes should be placed in the line after the last text "
    "in the docstring (do not close the quotes in the same line as "
    "the text, or leave a blank line between the last text and the "
    "quotes)",
    "GL03": "Double line break found; please use only one blank line to "
    "separate sections or paragraphs, and do not leave blank lines "
    "at the end of docstrings",
    "GL05": 'Tabs found at the start of line "{line_with_tabs}", please use '
    "whitespace only",
    "GL06": 'Found unknown section "{section}". Allowed sections are: '
    "{allowed_sections}",
    "GL07": "Sections are in the wrong order. Correct order is: {correct_sections}",
    "GL08": "The object does not have a docstring",
    "GL09": "Deprecation warning should precede extended summary",
    "GL10": "reST directives {directives} must be followed by two colons",
    "SS01": "No summary found (a short summary in a single line should be "
    "present at the beginning of the docstring)",
    "SS02": "Summary does not start with a capital letter",
    "SS03": "Summary does not end with a period",
    "SS04": "Summary contains heading whitespaces",
    "SS05": "Summary must start with infinitive verb, not third person "
    '(e.g. use "Generate" instead of "Generates")',
    "SS06": "Summary should fit in a single line",
    "ES01": "No extended summary found",
    "PR01": "Parameters {missing_params} not documented",
    "PR02": "Unknown parameters {unknown_params}",
    "PR03": "Wrong parameters order. Actual: {actual_params}. "
    "Documented: {documented_params}",
    "PR04": 'Parameter "{param_name}" has no type',
    "PR05": 'Parameter "{param_name}" type should not finish with "."',
    "PR06": 'Parameter "{param_name}" type should use "{right_type}" instead '
    'of "{wrong_type}"',
    "PR07": 'Parameter "{param_name}" has no description',
    "PR08": 'Parameter "{param_name}" description should start with a '
    "capital letter",
    "PR09": 'Parameter "{param_name}" description should finish with "."',
    "PR10": 'Parameter "{param_name}" requires a space before the colon '
    "separating the parameter name and type",
    "RT01": "No Returns section found",
    "RT02": "The first line of the Returns section should contain only the "
    "type, unless multiple values are being returned",
    "RT03": "Return value has no description",
    "RT04": "Return value description should start with a capital letter",
    "RT05": 'Return value description should finish with "."',
    "YD01": "No Yields section found",
    "SA01": "See Also section not found",
    "SA02": "Missing period at end of description for See Also "
    '"{reference_name}" reference',
    "SA03": "Description should be capitalized for See Also "
    '"{reference_name}" reference',
    "SA04": 'Missing description for See Also "{reference_name}" reference',
    "EX01": "No examples section found",
}
`
[warnings_cudf_numpydoc.txt](https://github.com/rapidsai/cudf/files/9603552/warnings_cudf_numpydoc.txt)

**Comment by vyasr** [comment]
Awesome! Here's a command that gives a quick idea of what warnings are showing up:
```
grep -v WARNING warnings_cudf_numpydoc.txt | sort | uniq
```

To get just a list of error codes you can also do
```
grep -v WARNING warnings_cudf_numpydoc.txt | sort | uniq | awk '{printf $1"\n"}' | uniq
```

From that, I see:
```
ES01:
EX01:
GL01:
GL02:
GL03:
GL06:
GL07:
GL08:
PR01:
PR02:
PR03:
PR04:
PR05:
PR06:
PR07:
PR08:
PR09:
PR10:
RT01:
RT02:
RT03:
RT04:
RT05:
SA01:
SA02:
SA04:
SS03:
SS05:
SS06:
*
/home/rishis/anaconda3/envs/cudf_dev/lib/python3.9/site-packages/cudf/core/series.py:docstring
```

From looking through that output, these all look like real errors that we should fix (although a couple of them are very pedantic like ending every line with a period). We can save the last few of lines until the end since those represent issues beyond just the docs (mostly `AttributeError` from something failing to import). The `numpydoc` website has [a full list of error codes](https://numpydoc.readthedocs.io/en/latest/validation.html#built-in-validation-checks) to explain what each of these are. Since these all do look like real errors, my recommendation would be to pick one or two of the above error codes at a time, find all of the issues, and make a PR fixing just that issue. That should make it fairly easy to get your PRs reviewed quickly. Does that make sense?

**Comment by ShahRishi** [comment]
This makes perfect sense. I will work on an error and create a pull request once I am ready. Thank you!

**Comment by vyasr** [comment]
@ShahRishi did you ever have any luck with this?

**Comment by ShahRishi** [comment]
Hi @vyasr. Thank you for reaching out, I unfortunately have had issues with this, so I slowed down. Is this something that is still an open issue?

**Comment by vyasr** [comment]
Hi @ShahRishi, yes it is and we would still welcome your contributions here. If there's something we can do to help unblock you, please let us know!

**Comment by vyasr** [comment]
@ShahRishi just circling back here. If you'd like, and if you have your code in some shareable state, feel free to merge the latest cudf and push out a draft PR! We can always get someone to help you out with finishing things up.

**Comment by ShahRishi** [comment]
Yes, I appreciate you checking in. I will consolidate my progress and push a draft by 5/17/24! Thank you @vyasr!

**Comment by vyasr** [comment]
Thank you @ShahRishi!

**Comment by pathak-satyam** [comment]
Hi, I'd like to pick this up. I'll start by enabling numpydoc validation and fixing a few error codes at a time.

**Comment by vyasr** [comment]
@pathak-satyam I apologize for not responding sooner. If you would like, please do!

**Comment by pathak-satyam** [comment]
Sure. I am on it
